### DISCUSSION TITLE
diesel_async and postgres pipelining

### DISCUSSION THREAD

**Discussion by art-in** [weiznich/diesel_async#4] [Q&A] [answered]
hey, i'm trying to use `diesel_async` with `actix_web` , but stuck on reusing same db connection for multiple http requests.

when using `tokio_postgres` directly:

```rs
pub struct Db {
    client: tokio_postgres::Client,
    find_user_statement: tokio_postgres::Statement,
}

pub async fn get_user(db: actix_web::web::Data<Db>, user_id: i32) -> Option<User> {
    let row = db
        .client
        .query_one(&db.find_user_statement, &[&user_id.clone()]) // it doesn't require client ref to be mutable
        .await
        .ok();

    row.map(|row| User {
        id: row.get(0),
        name: row.get(1),
    })
}
```

same with `diesel_async`:

```rs
pub struct Db {
    connection: diesel_async::AsyncPgConnection,
}

pub async fn get_user(
    db: actix_web::web::Data<Db>,
    user_id: i32,
) -> Result<models::User, diesel::result::Error> {
    use crate::schema::users::dsl::*;

    let user = users
        .filter(id.eq(user_id))
        .first::<models::User>(&mut db.connection) // error: expects mutable connection ref
        .await?;

    Ok(user)
}
```

i could wrap connection into async mutex or use connection pool, but that would neglect whole idea of reusing same connection for multiple queries simultaneously ([pipelining](https://docs.rs/tokio-postgres/latest/tokio_postgres/#pipelining))

so, is postgres pipelining possible with `diesel_async`?

**Comment by weiznich** [comment, marked as answer]
Pipelining is not supported yet. I cannot say if future versions will support pipelining as this requires quite a bit design work first to see if a common interface could abstract over the fact that a backend supports pipelining or not.

**Comment by Turbo87** [comment]
for anyone else finding this discussion in the future: at this point pipelining has officially been implemented at least for Postgres (see https://github.com/weiznich/diesel_async/blob/v0.5.0/src/pg/mod.rs#L850-L868)
